File: wiki-webapp/src/main/webapp/javascript/eXo/wiki/UIWikiTemplateForm.js
/**
 * Copyright (C) 2010 eXo Platform SAS.
 * 
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */(function(base, uiForm, webuiExt, $) {


function UIWikiTemplateForm() {
};

UIWikiTemplateForm.prototype.init = function(componentid, inputId, defaultText) {
  var me = eXo.wiki.UIWikiTemplateForm;
  me.component = document.getElementById(componentid);
  if (!me.component) {
    return;
  }
  var input = $(me.component).find('#' + inputId)[0];
  $(input).attr('autocomplete', 'off');
  $(input).keyup(function(evt) {
    evt = window.event || evt;
    eXo.wiki.UIWikiTemplateForm.pressHandler(evt, this);
  });
  eXo.wiki.UIWikiPortlet.decorateInput(input, defaultText, true);
  input.form.onsubmit = function() {
    return false;
  }
};

UIWikiTemplateForm.prototype.pressHandler = function(evt, textbox) {
  var me = eXo.wiki.UIWikiTemplateForm;
  evt = window.event || evt;
  var keyNum = eXo.wiki.UIWikiPortlet.getKeynum(evt);
  if (evt.altKey || evt.ctrlKey || evt.shiftKey)
    return;
  switch (keyNum) {
  case 13:
    me.enterHandler(evt);
    break;
  case 27:
    me.escapeHandler();
    break;
  case 38:
    me.arrowUpHandler();
    break;
  case 40:
    me.arrowDownHandler();
    break;
  default:
    break;
  }
  return;
};

/**
 * Handler key press
 */

UIWikiTemplateForm.prototype.enterHandler = function(evt){
    var me = eXo.wiki.UIWikiTemplateForm;
    me.doAdvanceSearch();
};

UIWikiTemplateForm.prototype.escapeHandler = function() {
};

UIWikiTemplateForm.prototype.arrowUpHandler = function() {
};

UIWikiTemplateForm.prototype.arrowDownHandler = function() {
};

UIWikiTemplateForm.prototype.typeHandler = function(textbox) {  
};

UIWikiTemplateForm.prototype.doAdvanceSearch = function() {
  var me = eXo.wiki.UIWikiTemplateForm;
  if (me.component) {
    var action = $(me.component).find('button.TemplateSearchButton')[0];
    action.onclick();
  }
}

eXo.wiki.UIWikiTemplateForm = new UIWikiTemplateForm();
return eXo.wiki.UIWikiTemplateForm;

})(base, uiForm, webuiExt, $);

File: wiki-webui/src/main/java/org/exoplatform/wiki/webui/commons/UIWikiTemplateForm.java
/*
 * Copyright (C) 2003-2010 eXo Platform SAS.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Affero General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see<http://www.gnu.org/licenses/>.
 */
package org.exoplatform.wiki.webui.commons;

import java.util.*;
import java.util.Map.Entry;

import org.exoplatform.commons.utils.LazyPageList;
import org.exoplatform.container.PortalContainer;
import org.exoplatform.webui.application.WebuiRequestContext;
import org.exoplatform.webui.config.annotation.ComponentConfig;
import org.exoplatform.webui.config.annotation.EventConfig;
import org.exoplatform.webui.core.UIGrid;
import org.exoplatform.webui.core.lifecycle.UIFormLifecycle;
import org.exoplatform.webui.event.Event;
import org.exoplatform.webui.event.EventListener;
import org.exoplatform.webui.form.UIForm;
import org.exoplatform.webui.form.UIFormStringInput;
import org.exoplatform.wiki.commons.Utils;
import org.exoplatform.wiki.mow.api.Template;
import org.exoplatform.wiki.service.WikiPageParams;
import org.exoplatform.wiki.service.WikiService;
import org.exoplatform.wiki.service.search.TemplateSearchData;
import org.exoplatform.wiki.service.search.TemplateSearchResult;
import org.exoplatform.wiki.webui.bean.TemplateBean;
import org.exoplatform.wiki.webui.bean.WikiTemplateListAccess;

@ComponentConfig(
  lifecycle = UIFormLifecycle.class,
  template = "app:/templates/wiki/webui/commons/UIWikiTemplateForm.gtmpl",
  events = {
    @EventConfig(listeners = UIWikiTemplateForm.SearchTemplateActionListener.class)
  })
public class UIWikiTemplateForm extends UIForm{
 
  public static final String    TEMPLATE_SEARCHBOX = "TemplateSeachBox";

  public static final String    TEMPLATE_GRID      = "UIWikiTemplateGrid";

  public static final String    TEMPLATE_ITER      = "TemplateIter";

  public static final int       ITEMS_PER_PAGE     = 8;

  public static final String    TEMPLATE_ID        = TemplateBean.ID;

  public static final String    TEMPLATE_NAME      = TemplateBean.TITLE;

  public static final String    DESCRIPTION        = TemplateBean.DESCRIPTION;

  public static final String[] TEMPLATE_FIELD     = { TEMPLATE_NAME, DESCRIPTION };

  public UIGrid                grid;

  public WikiService           wService;
  
  public ResourceBundle        res;
  
  public UIWikiTemplateForm() throws Exception {

    WebuiRequestContext context = WebuiRequestContext.getCurrentInstance();
    res = context.getApplicationResourceBundle();
    UIFormStringInput inputSearchBox = new UIFormStringInput(TEMPLATE_SEARCHBOX, null, null);
    inputSearchBox.setHTMLAttribute("placeholder", res.getString("UIWikiSearchBox.label.Search"));
    addChild(inputSearchBox);
    grid = addChild(UIWikiGrid.class, null, TEMPLATE_GRID);
    grid.getUIPageIterator().setId(TEMPLATE_ITER);
    grid.getUIPageIterator().setParent(this);
    wService = (WikiService) PortalContainer.getComponent(WikiService.class);
    ((UIWikiGrid)grid.configure(TEMPLATE_ID, TEMPLATE_FIELD, null)).setUIGridMode(UIWikiGrid.TEMPLATE);
    initGrid();
  }
  

  public void initGrid() throws Exception {
    List<TemplateBean> listBean = new ArrayList<>();

    WikiPageParams params = Utils.getCurrentWikiPageParams();
    Map<String, Template> templatesMap = wService.getTemplates(params);
    if(templatesMap != null) {
      for(Template template : templatesMap.values()) {
        listBean.add(new TemplateBean(template.getName(),
                template.getTitle(),
                template.getDescription()));
      }
    }
    LazyPageList<TemplateBean> lazylist = new LazyPageList<TemplateBean>(new WikiTemplateListAccess(listBean),
                                                                         ITEMS_PER_PAGE);
    grid.getUIPageIterator().setPageList(lazylist);
  }
  
  static public class SearchTemplateActionListener extends EventListener<UIWikiTemplateForm> {
    public void execute(Event<UIWikiTemplateForm> event) throws Exception {
      UIWikiTemplateForm form = event.getSource();
      WikiPageParams params = Utils.getCurrentWikiPageParams();
      UIFormStringInput searchbox = form.findComponentById(UIWikiTemplateForm.TEMPLATE_SEARCHBOX);
      String searchKeyword = searchbox.getValue();
      if (searchKeyword == null || (searchKeyword != null && searchKeyword.trim().length() == 0)) {
        form.initGrid();
      } else {
        TemplateSearchData data = new TemplateSearchData(searchKeyword, params.getType(), params.getOwner());
        List<TemplateSearchResult> results = form.wService.searchTemplate(data);
        List<TemplateBean> listBean = new ArrayList<TemplateBean>();
        for (int i = 0; i < results.size(); i++) {
          TemplateSearchResult result = results.get(i);
          listBean.add(new TemplateBean(result.getName(), result.getTitle(), result.getDescription()));
        }
        LazyPageList<TemplateBean> lazylist = new LazyPageList<TemplateBean>(new WikiTemplateListAccess(listBean), ITEMS_PER_PAGE);
        form.grid.getUIPageIterator().setPageList(lazylist);
      }
      event.getRequestContext().addUIComponentToUpdateByAjax(form);
    }
  }
}
